Editor/EditorCore/CutTool.cs

```diff
@@ -0,0 +1,1421 @@
+                        point = trs.TransformPoint(m_CutPath[m_SelectedIndex].position);
+
+                    float size = HandleUtility.GetHandleSize(point) * k_HandleSize;
+                    Handles.color = m_CurrentHandleColor;
+                    Handles.DotHandleCap(-1, point, Quaternion.identity, size, evt.type);
+                }
+                Handles.color = Color.white;
+            }
+        }
+
+        /// <summary>
+        /// Visual indications to help the user: highlighting faces, edges and vertices when snapping on them
+        /// </summary>
+        void DoVisualCues()
+        {
+            if(m_Mesh != null)
+            {
+                if(m_TargetFace == null)
+                {
+                    EditorHandleDrawing.HighlightVertices(m_Mesh, m_Mesh.sharedVertexLookup.Keys.ToArray(), false);
+                    EditorHandleDrawing.HighlightEdges(m_Mesh, m_Mesh.faces.SelectMany(f => f.edges).Distinct().ToArray(),
+                        false);
+
+                    if(m_CurrentFace != null)
+                        EditorHandleDrawing.HighlightFaces(m_Mesh, new Face[]{m_CurrentFace}, Color.Lerp(Color.blue, Color.cyan, 0.5f));
+                }
+                else
+                {
+                    var edges = m_TargetFace.edges;
+                    EditorHandleDrawing.HighlightVertices(m_Mesh, edges.Select(e => e.a).ToArray(), false);
+                    EditorHandleDrawing.HighlightEdges(m_Mesh, edges.ToArray(), false);
+
+                    if(m_SnapedVertexId != -1)
+                        EditorHandleDrawing.HighlightVertices(m_Mesh, new int[]{m_SnapedVertexId});
+
+                    if(m_SnapedEdge != Edge.Empty)
+                        EditorHandleDrawing.HighlightEdges(m_Mesh, new Edge[]{m_SnapedEdge});
+                }
+            }
+            else if(MeshSelection.activeMesh != null)
+            {
+                ProBuilderMesh mesh = MeshSelection.activeMesh;
+                EditorHandleDrawing.HighlightVertices(mesh, mesh.sharedVertexLookup.Keys.ToArray(), false);
+                EditorHandleDrawing.HighlightEdges(mesh, mesh.faces.SelectMany(f => f.edges).ToArray(), false);
+            }
+        }
+
+        /// <summary>
+        /// Rebuild the line mesh when updated
+        /// </summary>
+        void RebuildCutShape()
+        {
+            // If Undo is called immediately after creation this situation can occur
+            if (m_Mesh == null)
+                return;
+
+            UpdateMeshConnections();
+            ValidateCutShape();
+
+            // While the vertex count may not change, the triangle winding might. So unfortunately we can't take
+            // advantage of the `vertexCountChanged = false` optimization here.
+            ProBuilderEditor.Refresh();
+            SceneView.RepaintAll();
+        }
+
+        void ValidateCutShape()
+        {
+            Vertex[] vertices = m_Mesh.GetVertices();
```

Comment:
if this function is just reading the vertex data use `m_Mesh.positionsInternal` to skip the array copy.